### ISSUE TITLE
A pair of nodes in  a cluster becomes unresponsive: QueryPartitionOperation invocation failed to complete due to operation-heartbeat-timeout

### ISSUE THREAD

**Issue by ydekova** [hazelcast/hazelcast#19797]
We have a cluster of 15 nodes each on a separate VM in our data center:
We started having a problem almost every week - two of the nodes would "freeze" and become unresponsive.
The following error starts to occur on the client:

```
Address[10.128.41.218]:5701 has stopped heartbeating.
```
```
Target[Address[10.128.41.218]:5701] disconnected, connection was closed.
```
```
QueryPartitionOperation invocation failed to complete due to operation-heartbeat-timeout. Current time: 2021-10-15 19:22:34.064. Start time: 2021-10-15 19:21:33.066. Total elapsed time: 60998 ms. Last operation heartbeat: never. Last operation heartbeat from member: 2021-10-14 18:21:01.824. Invocation{op=com.hazelcast.map.impl.query.QueryPartitionOperation{serviceName='hz:impl:mapService', identityHash=734296934, partitionId=77, replicaIndex=0, callId=-28513230, invocationTime=1634325554910 (2021-10-15 19:19:14.910), waitTimeout=-1, callTimeout=60000, name=CoreTools.FixtureMatchDto}, tryCount=250, tryPauseMillis=500, invokeCount=1, callTimeoutMillis=60000, firstInvocationTimeMs=1634325693066, firstInvocationTime='2021-10-15 19:21:33.066', lastHeartbeatMillis=0, lastHeartbeatTime='1970-01-01 00:00:00.000', target=[10.128.41.218]:5701, pendingResponse={VOID}, backupsAcksExpected=0, backupsAcksReceived=0, connection=Connection[id=2, /10.128.41.221:49155->/10.128.41.218:5701, endpoint=[10.128.41.218]:5701, alive=true, type=MEMBER]}
```

Unfortunately we haven't been able to come up with a definite steps to reproduce it yet (working on it), since it happens on production randomly.
The issue is fixed after restart of the cluster or only the problematic nodes , but within two weeks it usually occurs again.
We have collected additional information listed below:

Hazelcast version: 3.8, build	20170217
java version 1.8, java VM version: 25.121-b13

Here are the properties of the nodes:

![hazel-prod-os](https://user-images.githubusercontent.com/8876741/137580055-0e56c2cc-d5f6-46c6-bf4f-19250efd8060.png)

![hazel-prod-memory](https://user-images.githubusercontent.com/8876741/137580061-5c42493f-34e1-447c-bc75-bc6297a58d2f.png)

The config file:

```xml
<hazelcast xsi:schemaLocation="http://www.hazelcast.com/schema/config hazelcast-config-3.8.xsd"
           xmlns="http://www.hazelcast.com/schema/config"
           xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
    <group>
        <name>CoreToolsPROD</name>
        <!-- <password>dev-pass</password> -->
    </group>
     <management-center enabled="true">http://vwp-crtoolswrk1.x.net:8081/mancenter/Record</management-center> 

    <network>
        <port auto-increment="true" port-count="100">5701</port>
        <outbound-ports>
            <!--
            Allowed port range when connecting to other nodes.
            0 or * means use system provided port.
            -->
            <ports>0</ports>
        </outbound-ports>
        <join>
            <multicast enabled="false">
                <multicast-group>224.2.2.3</multicast-group>
                <multicast-port>54327</multicast-port>
            </multicast>
            <tcp-ip enabled="true">
                <member-list>
                    <member>10.128.41.85-100</member>
                    <member>10.128.41.215-225</member>
                </member-list>
            </tcp-ip>
            <aws enabled="false">
                <access-key>my-access-key</access-key>
                <secret-key>my-secret-key</secret-key>
                <!--optional, default is us-east-1 -->
                <region>us-west-1</region>
                <!--optional, default is ec2.amazonaws.com. If set, region shouldn't be set as it will override this property -->
                <host-header>ec2.amazonaws.com</host-header>
                <!-- optional, only instances belonging to this group will be discovered, default will try all running instances -->
                <security-group-name>hazelcast-sg</security-group-name>
                <tag-key>type</tag-key>
                <tag-value>hz-nodes</tag-value>
            </aws>
            <discovery-strategies>
            </discovery-strategies>
        </join>
        <interfaces enabled="false">
            <interface>10.10.1.*</interface>
        </interfaces>
        <ssl enabled="false"/>
        <socket-interceptor enabled="false"/>
        <symmetric-encryption enabled="false">
            <!--
               encryption algorithm such as
               DES/ECB/PKCS5Padding,
               PBEWithMD5AndDES,
               AES/CBC/PKCS5Padding,
               Blowfish,
               DESede
            -->
            <algorithm>PBEWithMD5AndDES</algorithm>
            <!-- salt value to use when generating the secret key -->
            <salt>thesalt</salt>
            <!-- pass phrase to use when generating the secret key -->
            <password>thepass</password>
            <!-- iteration count to use when generating the secret key -->
            <iteration-count>19</iteration-count>
        </symmetric-encryption>
    </network>
    <partition-group enabled="false"/>
    <executor-service name="default">
        <pool-size>16</pool-size>
        <!--Queue capacity. 0 means Integer.MAX_VALUE.-->
        <queue-capacity>0</queue-capacity>
    </executor-service>
    <queue name="default">
        <!--
            Maximum size of the queue. When a JVM's local queue size reaches the maximum,
            all put/offer operations will get blocked until the queue size
            of the JVM goes down below the maximum.
            Any integer between 0 and Integer.MAX_VALUE. 0 means
            Integer.MAX_VALUE. Default is 0.
        -->
        <max-size>0</max-size>
        <!--
            Number of backups. If 1 is set as the backup-count for example,
            then all entries of the map will be copied to another JVM for
            fail-safety. 0 means no backup.
        -->
        <backup-count>0</backup-count>

        <!--
            Number of async backups. 0 means no backup.
        -->
        <async-backup-count>1</async-backup-count>

        <empty-queue-ttl>-1</empty-queue-ttl>
    </queue>
    <map name="default">
        <!--
           Data type that will be used for storing recordMap.
           Possible values:
           BINARY (default): keys and values will be stored as binary data
           OBJECT : values will be stored in their object forms
           NATIVE : values will be stored in non-heap region of JVM
        -->
        <in-memory-format>BINARY</in-memory-format>

        <!--
            Number of backups. If 1 is set as the backup-count for example,
            then all entries of the map will be copied to another JVM for
            fail-safety. 0 means no backup.
        -->
        <backup-count>1</backup-count>
        <!--
            Number of async backups. 0 means no backup.
        -->
        <async-backup-count>0</async-backup-count>
        <read-backup-data>true</read-backup-data>

        <!--
			Maximum number of seconds for each entry to stay in the map. Entries that are
			older than <time-to-live-seconds> and not updated for <time-to-live-seconds>
			will get automatically evicted from the map.
			Any integer between 0 and Integer.MAX_VALUE. 0 means infinite. Default is 0.
		-->
        <time-to-live-seconds>0</time-to-live-seconds>
        <!--
			Maximum number of seconds for each entry to stay idle in the map. Entries that are
			idle(not touched) for more than <max-idle-seconds> will get
			automatically evicted from the map. Entry is touched if get, put or containsKey is called.
			Any integer between 0 and Integer.MAX_VALUE. 0 means infinite. Default is 0.
		-->
        <max-idle-seconds>0</max-idle-seconds>
        <!--
            Valid values are:
            NONE (no eviction),
            LRU (Least Recently Used),
            LFU (Least Frequently Used).
            NONE is the default.
        -->
        <eviction-policy>NONE</eviction-policy>
        <!--
            Maximum size of the map. When max size is reached,
            map is evicted based on the policy defined.
            Any integer between 0 and Integer.MAX_VALUE. 0 means
            Integer.MAX_VALUE. Default is 0.
        -->
        <max-size policy="PER_NODE">0</max-size>
        <!--
            `eviction-percentage` property is deprecated and will be ignored when it is set.

            As of version 3.7, eviction mechanism changed.
            It uses a probabilistic algorithm based on sampling. Please see documentation for further details
        -->
        <eviction-percentage>25</eviction-percentage>
        <!--
            `min-eviction-check-millis` property is deprecated  and will be ignored when it is set.

            As of version 3.7, eviction mechanism changed.
            It uses a probabilistic algorithm based on sampling. Please see documentation for further details
        -->
        <min-eviction-check-millis>100</min-eviction-check-millis>
        <!--
            While recovering from split-brain (network partitioning),
            map entries in the small cluster will merge into the bigger cluster
            based on the policy set here. When an entry merge into the
            cluster, there might an existing entry with the same key already.
            Values of these entries might be different for that same key.
            Which value should be set for the key? Conflict is resolved by
            the policy set here. Default policy is PutIfAbsentMapMergePolicy

            There are built-in merge policies such as
            com.hazelcast.map.merge.PassThroughMergePolicy; entry will be overwritten if merging entry exists for the key.
            com.hazelcast.map.merge.PutIfAbsentMapMergePolicy ; entry will be added if the merging entry doesn't exist in the cluster.
            com.hazelcast.map.merge.HigherHitsMapMergePolicy ; entry with the higher hits wins.
            com.hazelcast.map.merge.LatestUpdateMapMergePolicy ; entry with the latest update wins.
        -->
        <merge-policy>com.hazelcast.map.merge.PutIfAbsentMapMergePolicy</merge-policy>

        <!--
           Control caching of de-serialized values. Caching makes query evaluation faster, but it cost memory.
           Possible Values:
                        NEVER: Never cache deserialized object
                        INDEX-ONLY: Caches values only when they are inserted into an index.
                        ALWAYS: Always cache deserialized values.
        -->
        <cache-deserialized-values>NEVER</cache-deserialized-values>

    </map>

    <multimap name="default">
        <backup-count>1</backup-count>
        <value-collection-type>SET</value-collection-type>
    </multimap>

    <list name="default">
        <backup-count>1</backup-count>
    </list>

    <set name="default">
        <backup-count>1</backup-count>
    </set>

<!--     <jobtracker name="default">
        <max-thread-size>0</max-thread-size>
        <queue-size>0</queue-size>
        <retry-count>0</retry-count>
        <chunk-size>1000</chunk-size>
        <communicate-stats>true</communicate-stats>
        <topology-changed-strategy>CANCEL_RUNNING_OPERATION</topology-changed-strategy>
    </jobtracker> -->

    <semaphore name="default">
        <initial-permits>0</initial-permits>
        <backup-count>1</backup-count>
        <async-backup-count>0</async-backup-count>
    </semaphore>

<!--     <reliable-topic name="default">
        <read-batch-size>10</read-batch-size>
        <topic-overload-policy>BLOCK</topic-overload-policy>
        <statistics-enabled>true</statistics-enabled>
    </reliable-topic> -->

    <ringbuffer name="default">
        <capacity>10000</capacity>
        <backup-count>1</backup-count>
        <async-backup-count>0</async-backup-count>
        <time-to-live-seconds>30</time-to-live-seconds>
        <in-memory-format>BINARY</in-memory-format>
    </ringbuffer>

    <properties>
        <property name="hazelcast.logging.type">log4j</property>
        <property name="hazelcast.event.thread.count">10</property>        
        <property name="hazelcast.event.queue.capacity">50000</property>        
        <property name="hazelcast.io.thread.count">50</property>        
        <property name="hazelcast.jmx">true</property>        
        <property name="hazelcast.jmx.detailed">true</property> 
        <property name="hazelcast.partition.count">3253</property>        
        <property name="hazelcast.operation.thread.count">500</property>
        <property name="hazelcast.client.shuffle.member.list">false</property>
    </properties>

    <services enable-defaults="true"/>

    <lite-member enabled="false"/>
</hazelcast>
```

We are starting the app with the following parameters:

`
START_ARGS=" -XX:+UseG1GC -XX:+UseStringDeduplication -XX:MaxGCPauseMillis=200 -XX:G1HeapRegionSize=4m"
START_ARGS=" -XX:ParallelGCThreads=16 -XX:ConcGCThreads=8"
START_ARGS="$START_ARGS -XX:+UseGCLogFileRotation -XX:NumberOfGCLogFiles=10 -XX:GCLogFileSize=50M -XX:+PrintGCDetails -XX:+PrintGCDateStamps -XX:+PrintStringDeduplicationStatistics -Xloggc:gc.log"
START_ARGS="$START_ARGS -Xmx26g -Xms26g"
START_ARGS="$START_ARGS -Dcom.sun.management.jmxremote -Dcom.sun.management.jmxremote.port=8303 -Dcom.sun.management.jmxremote.ssl=false -Dcom.sun.management.jmxremote.authenticate=false"
START_ARGS="$START_ARGS -Dlogback.configurationFile=file:/home/hazel/logback.xml"
START_ARGS="$START_ARGS -jar HazelCastModel-0.0.1.jar"
nohup java $START_ARGS -Xdebug -Xrunjdwp:transport=dt_socket,address=1044,server=y,suspend=n &
`

In the initialize method we are initializing 11 Maps, with BINARY storage. Most of the maps are used as simple dictionaries. Two maps have one index on a field for queries, and one  map has two indexes

```java
        IMap<String, FixtureMatchDto> fixtureMatchDtoMap = instance.getMap(Constants.MapNames.CoreTools.FixtureMatchDto);
        fixtureMatchDtoMap.addIndex("StartTime", true);
        fixtureMatchDtoMap.addIndex("FixtureId", false);
        log.info("CoreTools.FixtureMatchDto map is setup with index on StartTime");

```

The total heap cost of the maps is about 5GB
Most of the maps are small in size except for one, which  can reach 15 mil entities, that are frequently updated and and are indexed for queries.
The average put operation for this map is about 400 for a period of 10 secs, but might have spikes of up to 10 000/(10sec)

When the problem occurs the two hosts in question in question, in the example above those are:
10.128.41.218
10.128.41.221 
would continue to send metrics via JMX but would remain mostly unresponsive.
In the metrics there is no significant differences with the other hosts except for a sudden increase of the remote operations for those two nodes only (see in the image below a snapshot of the graphs for each server)

<img width="1041" alt="hazel-prod-metrics" src="https://user-images.githubusercontent.com/8876741/137581919-77ed8360-abed-4e4d-a627-06945e2658ed.png">

In the log files on the nodes we find the following errors;

```
2179013979 [hz._hzInstance_1_CoreToolsPROD.client-query.thread-5] ERROR c.h.map.impl.query.QueryOperation - [10.128.41.85]:5701 [CoreToolsPROD] [3.8] inconsistent range
java.lang.IllegalArgumentException: inconsistent range
	at java.util.concurrent.ConcurrentSkipListMap$SubMap.<init>(ConcurrentSkipListMap.java:2603) ~[na:1.8.0_121]
	at java.util.concurrent.ConcurrentSkipListMap.subMap(ConcurrentSkipListMap.java:2061) ~[na:1.8.0_121]
	at com.hazelcast.query.impl.SortedIndexStore.getSubRecordsBetween(SortedIndexStore.java:94) ~[HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.query.impl.IndexImpl.getSubRecordsBetween(IndexImpl.java:126) ~[HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.query.impl.predicates.BetweenPredicate.filter(BetweenPredicate.java:67) ~[HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.query.impl.predicates.AndPredicate.filter(AndPredicate.java:79) ~[HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.query.impl.Indexes.query(Indexes.java:112) ~[HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.map.impl.query.QueryRunner.runUsingIndexSafely(QueryRunner.java:120) ~[HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.map.impl.query.QueryRunner.runIndexOrPartitionScanQueryOnOwnedPartitions(QueryRunner.java:86) ~[HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.map.impl.query.QueryOperation.run(QueryOperation.java:48) ~[HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:186) ~[HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationExecutorImpl.run(OperationExecutorImpl.java:382) [HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationExecutorImpl.runOrExecute(OperationExecutorImpl.java:409) [HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.doInvokeLocal(Invocation.java:534) [HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.doInvoke(Invocation.java:519) [HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.invoke0(Invocation.java:490) [HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.invoke(Invocation.java:200) [HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.spi.impl.operationservice.impl.InvocationBuilderImpl.invoke(InvocationBuilderImpl.java:59) [HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.client.impl.protocol.task.map.AbstractMapQueryMessageTask.createInvocations(AbstractMapQueryMessageTask.java:124) [HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.client.impl.protocol.task.map.AbstractMapQueryMessageTask.invokeOnMembers(AbstractMapQueryMessageTask.java:101) [HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.client.impl.protocol.task.map.AbstractMapQueryMessageTask.call(AbstractMapQueryMessageTask.java:90) [HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.client.impl.protocol.task.AbstractCallableMessageTask.processMessage(AbstractCallableMessageTask.java:35) [HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.initializeAndProcessMessage(AbstractMessageTask.java:123) [HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.run(AbstractMessageTask.java:103) [HazelCastModel-0.0.1.jar:na]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_121]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_121]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_121]
	at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:76) [HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:92) [HazelCastModel-0.0.1.jar:na]
2179020238 [hz._hzInstance_1_CoreToolsPROD.generic-operation.thread-3] ERROR c.h.map.impl.query.QueryOperation - [10.128.41.85]:5701 [CoreToolsPROD] [3.8] inconsistent range
java.lang.IllegalArgumentException: inconsistent range
	at java.util.concurrent.ConcurrentSkipListMap$SubMap.<init>(ConcurrentSkipListMap.java:2603) ~[na:1.8.0_121]
	at java.util.concurrent.ConcurrentSkipListMap.subMap(ConcurrentSkipListMap.java:2061) ~[na:1.8.0_121]
	at com.hazelcast.query.impl.SortedIndexStore.getSubRecordsBetween(SortedIndexStore.java:94) ~[HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.query.impl.IndexImpl.getSubRecordsBetween(IndexImpl.java:126) ~[HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.query.impl.predicates.BetweenPredicate.filter(BetweenPredicate.java:67) ~[HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.query.impl.predicates.AndPredicate.filter(AndPredicate.java:79) ~[HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.query.impl.Indexes.query(Indexes.java:112) ~[HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.map.impl.query.QueryRunner.runUsingIndexSafely(QueryRunner.java:120) ~[HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.map.impl.query.QueryRunner.runIndexOrPartitionScanQueryOnOwnedPartitions(QueryRunner.java:86) ~[HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.map.impl.query.QueryOperation.run(QueryOperation.java:48) ~[HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:186) [HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:401) [HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:117) [HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.run(OperationThread.java:102) [HazelCastModel-0.0.1.jar:na]
2179025937 [hz._hzInstance_1_CoreToolsPROD.generic-operation.thread-1] ERROR c.h.map.impl.query.QueryOperation - [10.128.41.85]:5701 [CoreToolsPROD] [3.8] inconsistent range
java.lang.IllegalArgumentException: inconsistent range
	at java.util.concurrent.ConcurrentSkipListMap$SubMap.<init>(ConcurrentSkipListMap.java:2603) ~[na:1.8.0_121]
	at java.util.concurrent.ConcurrentSkipListMap.subMap(ConcurrentSkipListMap.java:2061) ~[na:1.8.0_121]
	at com.hazelcast.query.impl.SortedIndexStore.getSubRecordsBetween(SortedIndexStore.java:94) ~[HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.query.impl.IndexImpl.getSubRecordsBetween(IndexImpl.java:126) ~[HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.query.impl.predicates.BetweenPredicate.filter(BetweenPredicate.java:67) ~[HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.query.impl.predicates.AndPredicate.filter(AndPredicate.java:79) ~[HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.query.impl.Indexes.query(Indexes.java:112) ~[HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.map.impl.query.QueryRunner.runUsingIndexSafely(QueryRunner.java:120) ~[HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.map.impl.query.QueryRunner.runIndexOrPartitionScanQueryOnOwnedPartitions(QueryRunner.java:86) ~[HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.map.impl.query.QueryOperation.run(QueryOperation.java:48) ~[HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:186) [HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:401) [HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:117) [HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.run(OperationThread.java:102) [HazelCastModel-0.0.1.jar:na]
2179029292 [hz._hzInstance_1_CoreToolsPROD.generic-operation.thread-2] ERROR c.h.map.impl.query.QueryOperation - [10.128.41.85]:5701 [CoreToolsPROD] [3.8] inconsistent range
java.lang.IllegalArgumentException: inconsistent range
	at java.util.concurrent.ConcurrentSkipListMap$SubMap.<init>(ConcurrentSkipListMap.java:2603) ~[na:1.8.0_121]
	at java.util.concurrent.ConcurrentSkipListMap.subMap(ConcurrentSkipListMap.java:2061) ~[na:1.8.0_121]
	at com.hazelcast.query.impl.SortedIndexStore.getSubRecordsBetween(SortedIndexStore.java:94) ~[HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.query.impl.IndexImpl.getSubRecordsBetween(IndexImpl.java:126) ~[HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.query.impl.predicates.BetweenPredicate.filter(BetweenPredicate.java:67) ~[HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.query.impl.predicates.AndPredicate.filter(AndPredicate.java:79) ~[HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.query.impl.Indexes.query(Indexes.java:112) ~[HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.map.impl.query.QueryRunner.runUsingIndexSafely(QueryRunner.java:120) ~[HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.map.impl.query.QueryRunner.runIndexOrPartitionScanQueryOnOwnedPartitions(QueryRunner.java:86) ~[HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.map.impl.query.QueryOperation.run(QueryOperation.java:48) ~[HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:186) [HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:401) [HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:117) [HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.run(OperationThread.java:102) [HazelCastModel-0.0.1.jar:na]
2179046892 [hz._hzInstance_1_CoreToolsPROD.generic-operation.thread-2] ERROR c.h.map.impl.query.QueryOperation - [10.128.41.85]:5701 [CoreToolsPROD] [3.8] inconsistent range
java.lang.IllegalArgumentException: inconsistent range
	at java.util.concurrent.ConcurrentSkipListMap$SubMap.<init>(ConcurrentSkipListMap.java:2603) ~[na:1.8.0_121]
	at java.util.concurrent.ConcurrentSkipListMap.subMap(ConcurrentSkipListMap.java:2061) ~[na:1.8.0_121]
	at com.hazelcast.query.impl.SortedIndexStore.getSubRecordsBetween(SortedIndexStore.java:94) ~[HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.query.impl.IndexImpl.getSubRecordsBetween(IndexImpl.java:126) ~[HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.query.impl.predicates.BetweenPredicate.filter(BetweenPredicate.java:67) ~[HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.query.impl.predicates.AndPredicate.filter(AndPredicate.java:79) ~[HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.query.impl.Indexes.query(Indexes.java:112) ~[HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.map.impl.query.QueryRunner.runUsingIndexSafely(QueryRunner.java:120) ~[HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.map.impl.query.QueryRunner.runIndexOrPartitionScanQueryOnOwnedPartitions(QueryRunner.java:86) ~[HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.map.impl.query.QueryOperation.run(QueryOperation.java:48) ~[HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:186) [HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:401) [HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:117) [HazelCastModel-0.0.1.jar:na]
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.run(OperationThread.java:102) [HazelCastModel-0.0.1.jar:na]

```
Also, a lot of warnings:

`
2073176759 [hz._hzInstance_1_CoreToolsPROD.SlowOperationDetectorThread] WARN  c.h.s.i.o.s.SlowOperationDetector - [10.128.41.85]:5701 [CoreToolsPROD] [3.8] Slow operation detected: com.hazelcast.map.impl.operation.AddIndexOperation (4 invocations)`

It might be the case that the system needs scaling but it might be a bug or misconfiguration, so can you help identify what is currently wrong with our configuration?

**Comment by mmedenjak** [comment]
Thank you for the report @ydekova ! @vbekiaris @ahmetmircik can you take a look and provide some suggestions?

**Comment by vbekiaris** [comment]
hi @ydekova! Some intiial thoughts: a probable cause could be that under some circumstances you are issuing a query with  `BetweenPredicate` where `from` > `to` resulting in `IllegalArgumentException` being thrown from the internal sorted index implementation. Maybe you can add this kind of sanity check on your side to verify this could be the case.

Range sanity checks have been introduced in 3.12 with #14329, so it would be great if you could upgrade (Hazelcast 3.8 was released 4.5 years ago).
